Detectors/PHOS/calib/PHOSCalibWorkflow/src/PHOSEnergyCalibrator.cxx

```diff
@@ -0,0 +1,233 @@
+    }
+  }
+
+  //Real and Mixed inv mass distributions
+  // prepare TLorentsVector
+  float posX, posZ;
+  clu.getLocalPosition(posX, posZ);
+  TVector3 vec3;
+  mGeom->local2Global(clu.module(), posX, posZ, vec3);
+  vec3 -= mVertex;
+  float e = clu.getEnergy();
+  short absId;
+  mGeom->relPosToAbsId(clu.module(), posX, posZ, absId);
+
+  vec3 *= 1. / vec3.Mag();
+  TLorentzVector v(vec3.X() * e, vec3.Y() * e, vec3.Z() * e, e);
+  // Fill calibration histograms for all cells, even bad, but partners in inv, mass should be good
+  bool isGood = checkCluster(clu);
+  for (short ip = mBuffer->size(); ip--;) {
+    const TLorentzVector& vp = mBuffer->getEntry(ip);
+    TLorentzVector sum = v + vp;
+    if (mBuffer->isCurrentEvent(ip)) { //same (real) event
+      if (isGood) {
+        mHistos.fill(ETCalibHistos::kReInvMassNonlin, e, sum.M());
+      }
+      if (sum.Pt() > mPtMin) {
+        mHistos.fill(ETCalibHistos::kReInvMassPerCell, absId, sum.M());
+      }
+    } else { //Mixed
+      if (isGood) {
+        mHistos.fill(ETCalibHistos::kMiInvMassNonlin, e, sum.M());
+      }
+      if (sum.Pt() > mPtMin) {
+        mHistos.fill(ETCalibHistos::kMiInvMassPerCell, absId, sum.M());
+      }
+    }
+  }
+
+  //Add to list ot partners only if cluster is good
+  if (isGood) {
+    mBuffer->addEntry(v);
+  }
+}
+
+bool PHOSEnergySlot::checkCluster(const FullCluster& clu)
+{
+  //First check BadMap
+  float posX, posZ;
+  clu.getLocalPosition(posX, posZ);
+  short absId;
+  Geometry::relPosToAbsId(clu.module(), posX, posZ, absId);
+  if (!mBadMap->isChannelGood(absId)) {
+    return false;
+  }
+
+  return (clu.getEnergy() > 0.3 && clu.getMultiplicity() > 1);
+}
+
+//==================================================
+
+using es = o2::phos::PHOSEnergySlot;
+using Slot = o2::calibration::TimeSlot<o2::phos::PHOSEnergySlot>;
+PHOSEnergyCalibrator::PHOSEnergyCalibrator()
+{
+  // create final histos
+  mHistos.reset();
+}
+PHOSEnergySlot::PHOSEnergySlot(const PHOSEnergySlot& other)
+{
+  mRunStartTime = other.mRunStartTime;
+  mBuffer.reset(new RingBuffer());
+  mCalibParams.reset(new CalibParams(*(other.mCalibParams)));
+  mBadMap.reset(new BadChannelMap(*(other.mBadMap)));
+  mEvBC = other.mEvBC;
+  mEvOrbit = other.mEvOrbit;
+  mEvent = 0;
+  mPtMin = other.mPtMin;
+  mEminHGTime = other.mEminHGTime;
+  mEminLGTime = other.mEminLGTime;
+  mDigits.clear();
+  mHistos.reset();
+}
+
+void PHOSEnergyCalibrator::finalizeSlot(Slot& slot)
+{
+
+  // Extract results for the single slot
+  es* c = slot.getContainer();
+  LOG(INFO) << "Finalize slot " << slot.getTFStart() << " <= TF <= " << slot.getTFEnd();
+  //Add histos
+  mHistos.merge(c->getCollectedHistos());
+  //Add collected Digits
+  auto tmpD = c->getCollectedDigits();
+  //Add to list or write to file directly?
+  if (!mFout) { //not open yet?
+    LOG(INFO) << "Writing CalibDigits to file " << mdigitsfilename.data();
+    mFout.reset(TFile::Open(mdigitsfilename.data(), "recreate"));
+  }
```

Comment:
I see you are writing different slots to the same file. The calibrator will run trough up to the fill, would not it grow too much? Also, are you going to process (in calibration) the whole fill at once?

Reply:
what about this?